File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <map>
#include <algorithm>
#include <cmath>

using namespace std;

tuple<int, string, string> RecordParser(const string& InRecord)
{
    istringstream SS(InRecord);
    string Buffer;
    
    // 1. Time Parsing
    SS >> Buffer;
    int Pos = Buffer.find(':');
    int Hour = stoi(Buffer.substr(0, Pos));
    int Minute = stoi(Buffer.substr(Pos + 1));
    int Time = Hour * 60 + Minute;
    
    // 2. License Parsing
    SS >> Buffer;
    string License = Buffer;
    
    // 3. History Parsing
    SS >> Buffer;
    string History = Buffer;
    
    return {Time, License, History};
}

constexpr int LAST_EXIT = 1439;  // 23 * 60 + 59

tuple<int, int, int, int> FeeParser(const vector<int>& InFees)
{
    return {InFees[0], InFees[1], InFees[2], InFees[3]};
}

vector<int> solution(vector<int> InFees, vector<string> InRecords) 
{
    map<string, int> EntryMap, ParkingMap;
    for(const string& Record : InRecords)
    {
        const auto& [Time, License, History] = RecordParser(Record);
        if(History == "IN")
        {
            EntryMap[License] = Time;
            continue;
        }
        
        int EntryTime = EntryMap[License];
        EntryMap.erase(License);
        
        ParkingMap[License] += Time - EntryTime;
    }
    
    for(const auto& [License, EntryTime] : EntryMap)
    {
        ParkingMap[License] += LAST_EXIT - EntryTime;
    }
    
    const auto& [BaseTime, BaseFee, UnitTime, UnitFee] = FeeParser(InFees);
    vector<int> Answer(ParkingMap.size(), BaseFee);
    int Index = 0;
    for(auto& [License, Time] : ParkingMap)
    {
        if(Time > BaseTime)
        {
            Answer[Index] += ceil((double)(Time - BaseTime) / (double)UnitTime) * UnitFee;
        }
        ++Index;
    }
    return Answer;
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
from math import ceil

LAST_EXIT = 1439  # 24 * 60 + 59

def solution(in_fees, in_records):
    histories = {}
    for record in in_records:
        time, license, history = record.split()
        h, m = map(int, time.split(':'))
        time = h * 60 + m
        
        if license not in histories:
            histories[license] = []
        histories[license].append(time)
    
    parking = {}
    for license, history in histories.items():
        if len(history) & 1:
            history.append(LAST_EXIT)
        
        parking[license] = sum(exit - entry for entry, exit in zip(history[0::2], history[1::2]))
            
    base_time, base_fee, unit_time, unit_fee = in_fees
    answer = [base_fee] * len(parking)
    for i, (license, time) in enumerate(sorted(parking.items())):
        if time > base_time:
            answer[i] += ceil((time - base_time) / unit_time) * unit_fee
    
    return answer